patches/tock/12-pending-deadlock.patch

```diff
@@ -50,25 +52,38 @@ index 16b80cb10..c71ed639a 100644
 +            // The previous packet has not yet been received, reject the new one.
 +        } else {
 +            self.pending_out.set(true);
-+            // Process the first endpoint that has a delayed packet.
-+            // TODO(liamjm): Determine if a round-robin approach reduces latency
-+            // in the worst case.
-+            for (_, s) in self.endpoints.iter().enumerate() {
++            // Process the next endpoint that has a delayed packet.
++            for i in self.next_endpoint_index.get()..self.next_endpoint_index.get()+NUM_ENDPOINTS {
++                let s = &self.endpoints[i % NUM_ENDPOINTS];
                  // In case we reported Delay before, send the pending packet back to the client.
                  // Otherwise, there's nothing to do, the controller will send us a packet_out when a
                  // packet arrives.
-@@ -344,7 +348,7 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> ClientCtapHID<'a, 'b, C> {
+@@ -344,7 +349,7 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> ClientCtapHID<'a, 'b, C> {
                  .client
                  .map_or(false, |client| client.can_receive_packet(&app))
              {
 -                assert!(s.pending_out.take());
 +                assert!(self.pending_out.take());
- 
+
                  // Clear any pending packet on the transmitting side.
                  // It's up to the client to handle the received packet and decide if this packet
-@@ -387,8 +391,8 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> ClientCtapHID<'a, 'b, C> {
-     }
+@@ -352,6 +357,13 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> ClientCtapHID<'a, 'b, C> {
+                 self.cancel_in_transaction(endpoint);
  
+                 self.client.map(|client| client.packet_received(&buf, endpoint, app));
++                // Update next packet to send.
++                for (i, ep) in self.endpoints.iter().enumerate() {
++                  if ep.endpoint == endpoint {
++                    self.next_endpoint_index.set((i + 1) % NUM_ENDPOINTS);
++                    break
```

Comment:
nit: semi-colon after `break` (and other expressions used as statement that return `!`, i.e. any type, but don't rely on it, i.e. unit `()` would suffice).

Reply:
Done.